AWS Secrets Manager (#549)
hairyhenderson/gomplate: docs/content/datasources.md

| Type | URL Scheme(s) | Description |
|------|---------------|-------------|
| [AWS Systems Manager Parameter Store](#using-aws-smp-datasources) | `aws+smp` | [AWS Systems Manager Parameter Store][AWS SMP] is a hierarchically-organized key/value store which allows storage of text, lists, or encrypted secrets for retrieval by AWS resources |

> | [AWS Secrets Manager](#using-aws-secretsmanager-datasource) | `aws+asm` | [AWS Secrets Manager][] helps you protect secrets needed to access your applications, services, and IT resources. |

Reviewer: ```suggestion
| [AWS Secrets Manager](#using-aws-sm-datasource) | `aws+sm` | [AWS Secrets Manager][] helps you protect secrets needed to access your applications, services, and IT resources. |
```

## Using `aws+smp` datasources

> The `aws+smp://` scheme can be used to retrieve data from the [AWS Systems Manager](https://aws.amazon.com/systems-manager/) (née AWS EC2 Simple Systems Manager) [Parameter Store](https://aws.amazon.com/systems-manager/features/#Parameter_Store). This hierarchically organized key/value store allows you to store text, lists or encrypted secrets for easy retrieval by AWS resources. See [the AWS Systems Manager documentation](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-su-create.html#sysman-paramstore-su-create-about) for details on creating these parameters.

Author: Removed one whitespace: "AWS EC2 Simple  Systems Manager" > "AWS EC2 Simple Systems Manager"

aaa
```

## Using `aws+sm` datasource

### URL Considerations
For `aws+sm`, only the _scheme_ and _path_ components are necessary to be defined. Other URL components are ignored.

### Output

The output will be the SecretString from the `GetSecretValueOutput` object from the [AWS SDK for Go](https://docs.aws.amazon.com/sdk-for-go/api/service/secretsmanager/#GetSecretValueOutput)

### Examples

Given your [AWS account's Secret Manager](https://eu-central-1.console.aws.amazon.com/secretsmanager/home?region=eu-central-1#/listSecrets) has the following data:

> - `/foo/first/others` - `Bill,Ben` (a StringList)

Reviewer: This is incorrect - there's no `StringList` for this one

```suggestion
```

> - `/foo/first/password` - `super-secret` (a SecureString)

Reviewer: This is incorrect - there's no `SecureString` for this one

```suggestion
```

- `/foo/second/p1` - `aaa`

```console

> $ echo '{{ (ds "foo") }}' | gomplate -d foo=aws+sm:///foo/first/password

Reviewer: This example section needs to be revisited based on adjustments to the above data

aaa
```

> ## Using `aws+asm` datasource

Reviewer: ```suggestion
## Using `aws+sm` datasource
```

### URL Considerations

> For `aws+asm`, only the _scheme_ and _path_ components are necessary to be defined. Other URL components are ignored.

Reviewer: ```suggestion
For `aws+sm`, only the _scheme_ and _path_ components are necessary to be defined. Other URL components are ignored.
```

### Output

> The output will be a single `GetSecretValueOutput` object from the [AWS SDK for Go](https://docs.aws.amazon.com/sdk-for-go/api/service/secretsmanager/#GetSecretValueOutput)

Reviewer: > The output will be a single `GetSecretValueOutput` object

This behaviour has been quite awkward with the `aws+smp` datasource, and I totally regret it 😉. Can we avoid this and instead just return the `SecretString` string field? That way it's easier to substitute other datasources with the same template (imagine being able to run the same template with either a `vault` datasource or an `aws+sm` datasource, or a YAML `file` datasource for testing)...
